// zhlint@0.7.1/lib/options.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeConfig = exports.normalizeOptions = void 0;
const ignore_1 = __importStar(require("./hypers/ignore"));
const hexo_1 = __importDefault(require("./hypers/hexo"));
const vuepress_1 = __importDefault(require("./hypers/vuepress"));
const md_1 = __importDefault(require("./hypers/md"));
const rules_1 = require("./rules");
const report_1 = require("./report");
const hyperParseInfo = [
    { name: 'ignore', value: ignore_1.default },
    { name: 'hexo', value: hexo_1.default },
    { name: 'vuepress', value: vuepress_1.default },
    { name: 'markdown', value: md_1.default }
];
const arrToMap = (arr) => arr.reduce((current, { name, value }) => {
    current[name] = value;
    return current;
}, {});
const hyperParseMap = arrToMap(hyperParseInfo);
const matchCallArray = (calls, map) => calls
    .map((call) => {
    switch (typeof call) {
        case 'function':
            return call;
        case 'string':
            return map[call];
        default:
            return null;
    }
})
    .filter(Boolean);
const DEPRECATED_OPTIONS = {
    noSpaceInsideMark: 'noSpaceInsideWrapper',
    spaceBetweenHalfWidthContent: 'spaceBetweenHalfWidthLetters',
    noSpaceBetweenFullWidthContent: 'noSpaceBetweenFullWidthLetters',
    spaceBetweenMixedWidthContent: 'spaceBetweenMixedWidthLetters'
};
const deprecateOptions = (ruleOption, logger) => {
    var _a;
    for (const oldKey in DEPRECATED_OPTIONS) {
        const newKey = DEPRECATED_OPTIONS[oldKey];
        if (ruleOption[oldKey]) {
            logger.warn(`[deprecate] ${oldKey} is deprecated, use ${newKey} instead`);
            ruleOption[newKey] = (_a = ruleOption[newKey]) !== null && _a !== void 0 ? _a : ruleOption[oldKey];
            delete ruleOption[oldKey];
        }
    }
};
const normalizeOptions = (options) => {
    var _a, _b;
    const logger = (_a = options.logger) !== null && _a !== void 0 ? _a : report_1.env.defaultLogger;
    const rules = (_b = options.rules) !== null && _b !== void 0 ? _b : {};
    const preset = rules.preset === 'default' ? rules_1.defaultConfig : {};
    deprecateOptions(rules, logger);
    let hyperParse;
    if (typeof options.hyperParse === 'function') {
        hyperParse = [options.hyperParse];
    }
    else {
        hyperParse = options.hyperParse || hyperParseInfo.map((item) => item.name);
    }
    const normoalizedOptions = {
        logger,
        ignoredCases: options.ignoredCases || [],
        rules: Object.assign(Object.assign({}, preset), rules),
        hyperParse: matchCallArray(hyperParse, hyperParseMap)
    };
    return normoalizedOptions;
};
exports.normalizeOptions = normalizeOptions;
const normalizeConfig = (config, logger = report_1.env.defaultLogger) => {
    const options = {
        logger,
        rules: {},
        hyperParse: [],
        ignoredCases: []
    };
    let hyperParse = [];
    // preset
    if (config.preset === 'default') {
        options.rules = Object.assign({}, rules_1.defaultConfig);
        hyperParse = hyperParseInfo.map((item) => item.name);
    }
    // rules
    if (config.rules) {
        options.rules = Object.assign(Object.assign({}, options.rules), config.rules);
    }
    // hyper parsers
    if (Array.isArray(config.hyperParsers)) {
        hyperParse = config.hyperParsers;
    }
    hyperParse.forEach((x) => {
        if (!hyperParseMap[x]) {
            logger.log(`The hyper parser ${x} is invalid.`);
            return;
        }
        options.hyperParse.push(hyperParseMap[x]);
    });
    // ignored cases
    if (config.ignores) {
        config.ignores.forEach((x) => {
            const ignoredCase = (0, ignore_1.parseIngoredCase)(x);
            if (ignoredCase) {
                options.ignoredCases.push(ignoredCase);
            }
            else {
                logger.log(`The format of ignore case: "${x}" is invalid.`);
            }
        });
    }
    return options;
};
exports.normalizeConfig = normalizeConfig;
