// zhlint@0.7.1/lib/hypers/md.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const unified_1 = __importDefault(require("unified"));
const remark_parse_1 = __importDefault(require("remark-parse"));
const remark_frontmatter_1 = __importDefault(require("remark-frontmatter"));
const parser_1 = require("../parser");
const parsePosition = (position) => {
    var _a, _b;
    return ({
        start: ((_a = position === null || position === void 0 ? void 0 : position.start) === null || _a === void 0 ? void 0 : _a.offset) || 0,
        end: ((_b = position === null || position === void 0 ? void 0 : position.end) === null || _b === void 0 ? void 0 : _b.offset) || 0
    });
};
// AST related
const isParent = (node) => {
    return node.children !== undefined;
};
const blockTypes = ['paragraph', 'heading', 'table-cell'];
const isBlock = (node) => {
    return blockTypes.indexOf(node.type) >= 0;
};
const inlineContentTypes = [
    'emphasis',
    'strong',
    'delete',
    'footnote',
    'link',
    'linkReference'
];
const isInlineContent = (node) => {
    return inlineContentTypes.indexOf(node.type) >= 0;
};
const inlineRawTypes = [
    'inlineCode',
    'break',
    'image',
    'imageReference',
    'footnoteReference',
    'html'
];
const isInlineRaw = (node) => {
    return inlineRawTypes.indexOf(node.type) >= 0;
};
const travelBlocks = (node, blocks) => {
    if (isParent(node)) {
        node.children.forEach((child) => {
            if (child.type === 'yaml') {
                return;
            }
            if (isBlock(child)) {
                const blockMark = {
                    block: child,
                    inlineMarks: [],
                    hyperMarks: [],
                    value: '' // to be initialzed
                };
                blocks.push(blockMark);
                travelInlines(child, blockMark);
            }
            else {
                travelBlocks(child, blocks);
            }
        });
    }
};
const travelInlines = (node, blockMark) => {
    if (isParent(node)) {
        node.children.forEach((child) => {
            if (isInlineContent(child)) {
                blockMark.inlineMarks.push({ inline: child, raw: false });
                travelInlines(child, blockMark);
            }
            if (isInlineRaw(child)) {
                blockMark.inlineMarks.push({ inline: child, raw: true });
            }
        });
    }
};
const processBlockMark = (blockMark, str) => {
    const { block, inlineMarks } = blockMark;
    if (!block.position) {
        return;
    }
    const offset = block.position.start.offset || 0;
    const marks = [];
    const unresolvedCodeMarks = [];
    // Generate all the marks includes hyper (inline) and raw.
    inlineMarks.forEach((inlineMark) => {
        const { inline } = inlineMark;
        if (!inline.position) {
            return;
        }
        const startOffset = inline.position.start.offset || 0;
        const endOffset = inline.position.end.offset || 0;
        if (isInlineRaw(inline)) {
            const mark = {
                type: parser_1.MarkType.RAW,
                // TODO: typeof RawMark.meta
                meta: inline.type,
                startIndex: startOffset - offset,
                endIndex: endOffset - offset,
                startContent: str.substring(startOffset, endOffset),
                endContent: ''
            };
            // TODO: Ast.InlineCode?
            if (mark.startContent.match(/<code.*>/)) {
                const rawMark = Object.assign(Object.assign({}, mark), { code: parser_1.MarkSideType.LEFT });
                unresolvedCodeMarks.push(rawMark);
                marks.push(rawMark);
                return;
            }
            else if (mark.startContent.match(/<\/code.*>/)) {
                const rawMark = Object.assign(Object.assign({}, mark), { code: parser_1.MarkSideType.RIGHT });
                const leftCode = unresolvedCodeMarks.pop();
                if (leftCode) {
                    leftCode.rightPair = rawMark;
                }
                marks.push(rawMark);
                return;
            }
            marks.push(mark);
        }
        else {
            const firstChild = inline.children[0];
            const lastChild = inline.children[inline.children.length - 1];
            if (!firstChild.position || !lastChild.position) {
                return;
            }
            const innerStartOffset = firstChild.position.start.offset || 0;
            const innerEndOffset = lastChild.position.end.offset || 0;
            const mark = {
                type: parser_1.MarkType.HYPER,
                // TODO: typeof RawMark.meta
                meta: inline.type,
                startIndex: startOffset - offset,
                startContent: str.substring(startOffset, innerStartOffset),
                endIndex: innerEndOffset - offset,
                endContent: str.substring(innerEndOffset, endOffset)
            };
            marks.push(mark);
        }
    });
    blockMark.value = str.substring(block.position.start.offset || 0, block.position.end.offset || 0);
    blockMark.hyperMarks = marks
        .map((mark) => {
        if ((0, parser_1.isRawMark)(mark)) {
            if (mark.code === parser_1.MarkSideType.RIGHT) {
                return;
            }
            if (mark.code === parser_1.MarkSideType.LEFT) {
                const { rightPair } = mark;
                mark.startContent = str.substring(mark.startIndex + offset, mark.endIndex + offset);
                mark.endIndex = (rightPair === null || rightPair === void 0 ? void 0 : rightPair.endIndex) || 0;
                mark.endContent = '';
                delete mark.rightPair;
            }
        }
        return mark;
    })
        .filter(Boolean);
};
/**
  - travel all blocks/lists/tables/rows/cells
    - content: paragraph/heading/table-cell
    - no content: thematic break/code/html
  - for all phrasings:
    - no text: inline code/break/image/image ref/footnote ref/html
    - marks: emphasis/strong/delete/footnote/link/link ref
 */
const parser = (data) => {
    const content = data.content;
    const modifiedContent = data.modifiedContent;
    const ignoredByParsers = data.ignoredByParsers;
    const blockMarks = [];
    const tree = (0, unified_1.default)()
        .use(remark_parse_1.default)
        .use(remark_frontmatter_1.default)
        .parse(modifiedContent);
    // - travel and record all paragraphs/headings/table-cells into blocks
    // - for each block, travel and record all
    // - - 'hyper' marks: emphasis/strong/delete/footnote/link/linkRef and continue
    // - - 'raw' marks: inlineCode/break/image/imageRef/footnoteRef/html and stop
    travelBlocks(tree, blockMarks);
    // for each block marks
    // - get block.start.offset
    // - for each marks
    // - - startIndex: mark.start.offset - offset
    // - - startContent: [mark.start.offset - offset, mark.firstChild.start.offset - offset]
    // - - endIndex: mark.lastChild.end.offset - offset
    // - - endContent: [mark.lastChild.end.offset - offset, mark.end.offset]
    blockMarks.forEach((blockMark) => processBlockMark(blockMark, content));
    data.blocks = blockMarks.map((b) => {
        const position = parsePosition(b.block.position);
        ignoredByParsers.forEach(({ index, length, originContent: raw, meta }) => {
            if (position.start <= index && position.end >= index + length) {
                if (b.hyperMarks) {
                    b.hyperMarks.push({
                        type: parser_1.MarkType.RAW,
                        meta,
                        startIndex: index - position.start,
                        startContent: raw,
                        endIndex: index - position.start + length,
                        endContent: ''
                    });
                }
            }
        });
        return Object.assign({ value: b.value || '', marks: b.hyperMarks || [] }, position);
    });
    data.ignoredByParsers = [];
    return data;
};
exports.default = parser;
