// zhlint@0.7.1/lib/rules/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = void 0;
const space_trim_1 = __importDefault(require("./space-trim"));
const punctuation_width_1 = __importDefault(require("./punctuation-width"));
const punctuation_unification_1 = __importDefault(require("./punctuation-unification"));
const case_abbrs_1 = __importDefault(require("./case-abbrs"));
const space_wrapper_1 = __importDefault(require("./space-wrapper"));
const space_code_1 = __importDefault(require("./space-code"));
const space_letters_1 = __importDefault(require("./space-letters"));
const space_punctuation_1 = __importDefault(require("./space-punctuation"));
const space_quote_1 = __importDefault(require("./space-quote"));
const space_bracket_1 = __importDefault(require("./space-bracket"));
const case_linebreak_1 = __importDefault(require("./case-linebreak"));
const case_zh_units_1 = __importDefault(require("./case-zh-units"));
const case_html_entity_1 = __importDefault(require("./case-html-entity"));
const generateHandlers = (options) => {
    return [
        (0, space_trim_1.default)(options),
        (0, punctuation_width_1.default)(options),
        (0, punctuation_unification_1.default)(options),
        (0, case_abbrs_1.default)(options),
        (0, space_wrapper_1.default)(options),
        (0, space_code_1.default)(options),
        (0, space_letters_1.default)(options),
        (0, space_punctuation_1.default)(options),
        (0, space_quote_1.default)(options),
        (0, space_bracket_1.default)(options),
        (0, case_linebreak_1.default)(options),
        (0, case_zh_units_1.default)(options),
        (0, case_html_entity_1.default)(options)
    ];
};
exports.defaultConfig = {
    noSinglePair: true,
    halfWidthPunctuation: `()`,
    fullWidthPunctuation: `，。：；？！“”‘’`,
    adjustedFullWidthPunctuation: `“”‘’`,
    unifiedPunctuation: 'simplified',
    spaceBetweenHalfWidthLetters: true,
    noSpaceBetweenFullWidthLetters: true,
    spaceBetweenMixedWidthLetters: true,
    noSpaceBeforePunctuation: true,
    spaceAfterHalfWidthPunctuation: true,
    noSpaceAfterFullWidthPunctuation: true,
    spaceOutsideHalfQuote: true,
    noSpaceOutsideFullQuote: true,
    noSpaceInsideQuote: true,
    spaceOutsideHalfBracket: true,
    noSpaceOutsideFullBracket: true,
    noSpaceInsideBracket: true,
    spaceOutsideCode: true,
    noSpaceInsideWrapper: true,
    trimSpace: true,
    skipZhUnits: `年月日天号时分秒`,
    skipAbbrs: [
        'Mr.',
        'Mrs.',
        'Dr.',
        'Jr.',
        'Sr.',
        'vs.',
        'etc.',
        'i.e.',
        'e.g.',
        'a.k.a.'
    ]
};
exports.default = generateHandlers;
