// zhlint@0.7.1/lib/rules/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRIM_SPACE = exports.QUOTE_SPACE_OUTSIDE = exports.QUOTE_NOSPACE_OUTSIDE = exports.QUOTE_NOSPACE_INSIDE = exports.CONTENT_NOSPACE_MIXED_WIDTH = exports.CONTENT_SPACE_MIXED_WIDTH = exports.CONTENT_NOSPACE_FULL_WIDTH = exports.CONTENT_SPACE_HALF_WIDTH = exports.BRACKET_SPACE_OUTSIDE = exports.BRACKET_NOSPACE_OUTSIDE = exports.BRACKET_NOSPACE_INSIDE = exports.PUNCTUATION_SPACE_AFTER = exports.PUNCTUATION_NOSPACE_AFTER = exports.PUNCTUATION_NOSPACE_BEFORE = exports.PUNCTUATION_HALF_WIDTH = exports.PUNCTUATION_FULL_WIDTH = exports.PUNCTUATION_UNIFICATION_SIMPLIFIED = exports.PUNCTUATION_UNIFICATION_TRADITIONAL = exports.MARKDOWN_NOSPACE_INSIDE = exports.MARKDOWN_SPACE_OUTSIDE = exports.CODE_NOSPACE_OUTSIDE = exports.CODE_SPACE_OUTSIDE = void 0;
exports.CODE_SPACE_OUTSIDE = '此处内联代码的外部需要一个空格';
exports.CODE_NOSPACE_OUTSIDE = '此处内联代码的外部不需要空格';
exports.MARKDOWN_SPACE_OUTSIDE = '此处 Markdown 标记的空格需要在外部';
exports.MARKDOWN_NOSPACE_INSIDE = '此处 Markdown 标记的内部不需要空格';
exports.PUNCTUATION_UNIFICATION_TRADITIONAL = '此处标点符号需要统一到繁体';
exports.PUNCTUATION_UNIFICATION_SIMPLIFIED = '此处标点符号需要统一到简体';
exports.PUNCTUATION_FULL_WIDTH = '此处标点符号需要使用全角';
exports.PUNCTUATION_HALF_WIDTH = '此处标点符号需要使用半角';
exports.PUNCTUATION_NOSPACE_BEFORE = '此处标点符号前不需要空格';
exports.PUNCTUATION_NOSPACE_AFTER = '此处标点符号后不需要空格';
exports.PUNCTUATION_SPACE_AFTER = '此处标点符号后需要一个空格';
exports.BRACKET_NOSPACE_INSIDE = '此处括号的内部不需要空格';
exports.BRACKET_NOSPACE_OUTSIDE = '此处括号的外部不需要空格';
exports.BRACKET_SPACE_OUTSIDE = '此处括号的外部需要一个空格';
exports.CONTENT_SPACE_HALF_WIDTH = '此处半角内容之间需要一个空格';
exports.CONTENT_NOSPACE_FULL_WIDTH = '此处全角内容之间不需要空格';
exports.CONTENT_SPACE_MIXED_WIDTH = '此处中英文内容之间需要一个空格';
exports.CONTENT_NOSPACE_MIXED_WIDTH = '此处中英文内容之间需要一个空格';
exports.QUOTE_NOSPACE_INSIDE = '此处引号的内部不需要空格';
exports.QUOTE_NOSPACE_OUTSIDE = '此处引号的外部不需要空格';
exports.QUOTE_SPACE_OUTSIDE = '此处引号的外部需要一个空格';
exports.TRIM_SPACE = '此处需要去除外部空格';
